Wrap 2.0

JohnSundell opened the pull request:
This PR is the proposal for a 2.0 release of Wrap, which is the release that will bring `master` onto Swift 3 (compatibility with Swift 2.x will be kept through the `1.1.1` release). It contains the contents of the `swift3` branch, as well as some API refinements, improvement suggestions & fixes for opened issues:
- `Wrap(..)` is now `wrap(..)`
- `Wrap(objects:)` is now `wrap()`
- `keyForWrapping(propertyName:)` is now `keyForWrapping(propertyNamed:)`
- `WrapCustomizable.wrap()` is now `wrap(context:dateFormatter:)`
- `wrap(propertyName:originalValue:)` is now `wrap(propertyNamed:originalValue:context:dateFormatter:)`
- All enums use a lower case leading character
- Wrapping to `snake_case` is now supported (`myProperty ->`my_property`)
- Int64 & UInt64 are now encoded properly on 32 bit systems
- Date Formatters are now kept througout the entire wrapping process, and can be used in customized wrapping
- Wrap now has support for contextual objects just like Unbox has (send `context:`) when initiating the wrapping

JohnSundell committed:
Support Swift 3

JohnSundell committed:
Convert to latest Swift 3 syntax + add support for native Date type

JohnSundell committed:
Update for Xcode 8 beta 3

JohnSundell committed:
Correct optional binding in test

JohnSundell committed:
Update for Xcode 8 beta 5

JohnSundell committed:
Update for Xcode 8 release

JohnSundell committed:
Remove argument label for Wrap(objects:)

JohnSundell committed:
Use propertyNamed propertyName argument labelling to be Swift 3 compliant

JohnSundell committed:
Merge branch 'master' into 2.0

JohnSundell committed:
Use lower-case leading character for WrapError enum

JohnSundell committed:
Use lower-case leading character for VerificationError enum

JohnSundell committed:
Fix documentation error due to enum name

JohnSundell committed:
Wrap() -> wrap() to follow Swift 3 API naming conventions

JohnSundell committed:
Make all test enum cases lowercase

JohnSundell committed:
Update README for Swift 3 APIs and names

JohnSundell committed:
Merge branch 'master' into 2.0

JohnSundell committed:
Fix conflicts after merge with master

JohnSundell committed:
Verify(..) -> verify(..)

JohnSundell committed:
Travis: Use xcodebuild instead of xctool

JohnSundell committed:
Set SWIFT_VERSION = 3.0 for all targets

JohnSundell committed:
Enable running tests for tvOS from Xcode

JohnSundell committed:
Merge branch 'master' into 2.0

awclives commented:
Sounds reasonable to me.   One issue that is probably worth looking into is Int64 support.  The objective c bridgeable trick mentioned here: https://github.com/JohnSundell/Wrap/issues/13 does not work with Swift 3.0.

I think this might work with your new code.  Haven't been able to test fully yet though.

```
extension Int64: WrapCustomizable {
    public func wrap() -> Any? {
        return NSNumber(value: self)
    }
}

```

awclives commented:
Have done further testing, and the code in my previous comment seems to be working well.  Without it, if you include a Int64 in your object, the resulting wrap dictionary will cause the Foundation JSON serialization code to fail because it does not recognize Int64.

https://developer.apple.com/reference/foundation/jsonserialization

I think the other numbers are covered because they are objectivecbridgeable.

JohnSundell commented:
@awclives That's awesome, will add it onto this PR!

JohnSundell committed:
Ensure Int64 & UInt64 gets wrapped properly by using an NSNumber wrapper

JohnSundell committed:
Tests: Avoid infinite recursion when Obj-C types were mismatched

JohnSundell committed:
Pass supplied DateFormatter to WrapCustomizable objects

Resolves https://github.com/JohnSundell/Wrap/issues/10

JohnSundell committed:
Add option to convert keys to snake_case

Thanks to @johannth for providing first implementation in
https://github.com/JohnSundell/Wrap/pull/14

JohnSundell committed:
Add support for contextual objects when Wrapping

Works just like Unbox, supply a context of Any type, and it’ll be available
throughout the wrapping process.

JohnSundell committed:
README: Update method signature and add snake_case documentation

JohnSundell committed:
README: Document contextual objects

JohnSundell committed:
Update to 2.0 for Podspec & Swift Package Manager recommendation

JohnSundell committed:
README: Add note on compatibility

JohnSundell commented:
OK let's 🚢 this! 🎉
